File: db/scrape.js
#!/usr/bin/env node

/*
 * Copyright (c) 2015 Alex Yatskov <alex@foosoft.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

var cheerio    = require('cheerio');
var request    = require('request');
var underscore = require('underscore');
var url        = require('url');
var path       = require('path');
var fs         = require('fs');
var _          = require('underscore');


function requestCached(relativeUrl, callback) {
    var absoluteUrl = url.resolve('http://www.tripadvisor.com', relativeUrl);
    var cachePath   = path.join('cache', relativeUrl);

    fs.readFile(cachePath, function(err, data) {
        if (err) {
            var stream = fs.createWriteStream(cachePath);
            request(absoluteUrl, callback).pipe(stream);
        }
        else {
            callback(null, null, data);
        }
    });
}

function getBarPercent(bar) {
    return parseFloat(bar.attr('alt')) / 5.0;
}

function reviewScraped(err, resp, html) {
    if (err) {
        return console.error('Error: %s', err);
    }

    var $ = cheerio.load(html);

    var address = $('address span.format_address').text().trim();
    if (!address) {
        console.warn('Warning: review skipped, no address');
        return;
    }

    var storeName = $('h1#HEADING').text().trim();
    if (storeName.indexOf('CLOSED') !== -1) {
        console.warn('Warning: review skipped, closed');
        return;
    }

    var rating = $('ul.barChart img.rating_s_fill');
    if (rating.length != 4) {
        console.warn('Warning: review skipped, no summary');
        return;
    }

    var rateFood       = getBarPercent($(rating[0]));
    var rateService    = getBarPercent($(rating[1]));
    var rateValue      = getBarPercent($(rating[2]));
    var rateAtmosphere = getBarPercent($(rating[3]));

    if (rateFood === 0.0 && rateService === 0.0 && rateValue === 0.0 && rateAtmosphere === 0.0) {
        console.warn('Warning: review skipped, empty review');
        return;
    }

    var data = {
        name:        storeName,
        relativeUrl: this.relativeUrl,
        address:     address,
        rating: {
            food:       (rateFood - 0.5) * 2.0,
            service:    (rateService - 0.5) * 2.0,
            value:      (rateValue - 0.5) * 2.0,
            atmosphere: (rateAtmosphere - 0.5) * 2.0
        }
    };

    this.callback(data);
}

function scrapeReview(relativeUrl, callback) {
    console.log('Scraping review %s...', relativeUrl);

    var c = _.bind(reviewScraped, {
        callback:    callback,
        relativeUrl: relativeUrl
    });
    requestCached(relativeUrl, c);
}

function indexScraped(err, resp, html) {
    if (err) {
        return console.error('Error: %s', err);
    }

    var $     = cheerio.load(html);
    var that  = this;
    var abort = false;

    $('a.property_title').each(function(index, element) {
        if (abort) {
            return;
        }

        var reviewUrl = $(element).attr('href');
        if (that.callback(reviewUrl)) {
            abort = true;
        }
    });

    if (!abort) {
        var nextPageUrl = $('a.sprite-pageNext').attr('href');
        if (nextPageUrl) {
            scrapeIndices(nextPageUrl, this.callback);
        }
    }
}

function scrapeIndices(relativeUrl, callback) {
    console.log('Scraping index %s...', relativeUrl);

    var c = _.bind(indexScraped, { callback: callback });
    requestCached(relativeUrl, c);
}

function main() {
    var relativePaths = [
        '/Restaurants-g298173-Yokohama_Kanagawa_Prefecture_Kanto.html',
        '/Restaurants-g1021277-Fujisawa_Kanagawa_Prefecture_Kanto.html',
        '/Restaurants-g1021279-Chigasaki_Kanagawa_Prefecture_Kanto.html',
        '/Restaurants-g298172-Kawasaki_Kanagawa_Prefecture_Kanto.html',
        '/Restaurants-g1066854-Shinagawa_Tokyo_Tokyo_Prefecture_Kanto.html',
        '/Restaurants-g298184-Tokyo_Tokyo_Prefecture_Kanto.html'
    ];
    var databasePath = 'data.json';

    var abort = false;
    process.on('SIGINT', function() {
        console.warn('Caught SIGINT, aborting...');
        abort = true;
    });

    var results = [];
    _.each(relativePaths, function(relativePath) {
        scrapeIndices(relativePath, function(relativeUrl) {
            scrapeReview(relativeUrl, function(data) {
                results.push(data);
            });

            return abort;
        });
    });

    process.on('exit', function() {
        console.log('Total reviews scraped: %d', results.length);
        var strData = JSON.stringify(results, null, 4);
        fs.writeFileSync(databasePath, strData);
    });
}


if (require.main === module) {
    main();
}

File: scrape.go
/*
 * Copyright (c) 2015 Alex Yatskov <alex@foosoft.net>
 * Author: Alex Yatskov <alex@foosoft.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package main

import (
	"bufio"
	"database/sql"
	"errors"
	"net/url"
	"os"

	_ "github.com/mattn/go-sqlite3"
)

func scrapeUrls(filename string, wc *webCache, gc *geoCache) ([]restaurant, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var results []restaurant
	var scanner = bufio.NewScanner(file)

	for scanner.Scan() {
		if line := scanner.Text(); len(line) > 0 {
			parsed, err := url.Parse(line)
			if err != nil {
				return nil, err
			}

			var items []restaurant
			switch parsed.Host {
			case "tabelog.com":
				items = scrape(line, wc, gc, tabelog{})
			case "www.tripadvisor.com":
				items = scrape(line, wc, gc, tripadvisor{})
			default:
				return nil, errors.New("unsupported review site")
			}

			results = append(results, items...)
		}
	}

	return results, nil
}

func scrapeData(urlsPath, geocachePath, webcachePath string) ([]restaurant, error) {
	gc, err := newGeoCache(geocachePath)
	if err != nil {
		return nil, err
	}
	defer gc.save()

	wc, err := newWebCache(webcachePath)
	if err != nil {
		return nil, err
	}

	restaurants, err := scrapeUrls(urlsPath, wc, gc)
	if err != nil {
		return nil, err
	}

	return restaurants, nil
}

func processData(restaurants []restaurant, stationsPath string) error {
	sq, err := newStationQuery(stationsPath)
	if err != nil {
		return err
	}

	for i, _ := range restaurants {
		r := &restaurants[i]
		r.closestStnName, r.closestStnDist = sq.closestStation(r.latitude, r.longitude)
	}

	return nil
}

func buildFeatures(r restaurant) (delicious, accommodating, affordable, atmospheric float64) {
	return r.features["food"], r.features["service"], r.features["value"], r.features["atmosphere"]
}

func dumpData(dbPath string, restaraunts []restaurant) error {
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return err
	}
	defer db.Close()

	//
	// Restaurants
	//

	_, err = db.Exec(`
		DROP TABLE IF EXISTS reviews;
		CREATE TABLE reviews(
			name VARCHAR(100) NOT NULL,
			url VARCHAR(200) NOT NULL,
			delicious FLOAT NOT NULL,
			accommodating FLOAT NOT NULL,
			affordable FLOAT NOT NULL,
			atmospheric FLOAT NOT NULL,
			latitude FLOAT NOT NULL,
			longitude FLOAT NOT NULL,
			closestStnDist FLOAT NOT NULL,
			closestStnName VARCHAR(100) NOT NULL,
			accessCount INT NOT NULL,
			id INT PRIMARY KEY
		)`)

	if err != nil {
		return err
	}

	for _, r := range restaraunts {
		delicious, accommodating, affordable, atmospheric := buildFeatures(r)

		_, err = db.Exec(`
			INSERT INTO reviews(
				name,
				url,
				delicious,
				accommodating,
				affordable,
				atmospheric,
				latitude,
				longitude,
				closestStnDist,
				closestStnName,
				accessCount
			) VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
			r.name,
			r.url,
			delicious,
			accommodating,
			affordable,
			atmospheric,
			r.longitude,
			r.latitude,
			r.closestStnDist,
			r.closestStnName,
			0)

		if err != nil {
			return err
		}
	}

	//
	// Categories
	//

	_, err = db.Exec(`
		DROP TABLE IF EXISTS categories;
		CREATE TABLE categories(
			description VARCHAR(200) NOT NULL,
			id INT PRIMARY KEY)`)

	if err != nil {
		return err
	}

	for _, category := range []string{"I prefer quiet places", "I enjoy Mexican Food", "I drive a car"} {
		if _, err := db.Exec("INSERT INTO categories(description) VALUES (?)", category); err != nil {
			return err
		}
	}

	//
	// History
	//

	_, err = db.Exec(`
		DROP TABLE IF EXISTS history;
		CREATE TABLE history(
			date DATETIME NOT NULL,
			reviewId INT NOT NULL,
			id INT PRIMARY KEY,
			FOREIGN KEY(reviewId) REFERENCES reviews(id))`)

	if err != nil {
		return err
	}

	//
	// HistoryGroup
	//

	_, err = db.Exec(`
		DROP TABLE IF EXISTS historyGroups;
		CREATE TABLE historyGroups(
			categoryId INT NOT NULL,
			categoryValue FLOAT NOT NULL,
			historyId INT NOT NULL,
			FOREIGN KEY(historyId) REFERENCES history(id),
			FOREIGN KEY(categoryId) REFERENCES categories(id))`)

	if err != nil {
		return err
	}

	return nil
}

func main() {
	restaurants, err := scrapeData("data/urls.txt", "cache/geocache.json", "cache/webcache")
	if err != nil {
		panic(err)
	}

	if err := processData(restaurants, "data/stations.json"); err != nil {
		panic(err)
	}

	if err := dumpData("db.sqlite3", restaurants); err != nil {
		panic(err)
	}
}
